pkg/arvo/app/thread-watch-hook.hoon

```diff
@@ -0,0 +1,199 @@
+::
+++  on-init
+  |^  ^-  (quip card _this)
+  :_  this
+  :_  ~
+  (act /inv-gra [%watch %invite-store /invitatory/graph %invite-accepted-graph])
+  ::
+  ++  act
+    |=  [=wire =action:sur]
+    ^-  card
+    :*  %pass
+        wire
+        %agent
+        [our.bowl %thread-watch-hook]
+        %poke
+        %thread-watch-action
+        !>  ^-  action:sur
+        action
+    ==
+  --
+::
+++  on-save   !>(state)
+++  on-load
+  |=  old-vase=vase
+  ^-  (quip card _this)
+  `this(state !<(state-0 old-vase))
+::
+++  on-poke
+  |=  [=mark =vase]
+  ^-  (quip card _this)
+  ?>  (team:title our.bowl src.bowl)
+  ?.  ?=(%thread-watch-action mark)
+    (on-poke:def mark vase)
+  =/  =action:sur  !<(action:sur vase)
+  =*  observer  observer.action
+  =/  vals  (silt ~(val by observers))
+  ?-  -.action
+      %watch
+    ?:  ?|(=(app.observer %spider) =(app.observer %thread-watch-hook)) 
+      ~|('we avoid infinite loops' !!)
+    ?:  (~(has in vals) observer)
+      ~|('duplicate observer' !!)
+    :_  this(observers (~(put by observers) now.bowl observer))  
+    :_  ~
+    :*  %pass
+        /observer/(scot %da now.bowl)
+        %agent
+        [our.bowl app.observer]
+        %watch
+        path.observer
+    == 
+  ::
+      %ignore
+    ?.  (~(has in vals) observer)
+      ~|('cannot remove nonexistent observer' !!)
+    =/  key  (got-by-val observers observer)
+    :_  this(observers (~(del by observers) key))
+    :_  ~
+    :*  %pass
+        /observer/(scot %da key)
+        %agent
+        [our.bowl app.observer]
+        %leave
+        ~
+    == 
+  ==
+::
+++  on-agent
+  |=  [=wire =sign:agent:gall]
+  ^-  (quip card _this)
+  ?-    -.sign
+      %kick
+    ?+    wire  (on-agent:def wire sign)
+        [%thread-result @ ~]  [~ this]
+        [%observer @ ~]
+      =/  =time  (slav %da i.t.wire)
+      =/  =observer:sur  (~(got by observers) time)
+      :_  this
+      :_  ~
+      :*  %pass
+          wire
+          %agent
+          [our.bowl app.observer]
+          %watch
+          path.observer
+      == 
+    ==
+  ::
+      %poke-ack
+    ?.  ?=([%thread-start @ @ ~] wire)
+      (on-agent:def wire sign)
+    ?~  p.sign  [~ this]
+    =/  =time  (slav %da i.t.wire)
+    =/  tid  (slav %uv i.t.t.wire)
+    =/  =observer:sur  (~(got by observers) time)
+    :_  this(observers (~(del by observers) time))
+    :~  :*  %pass
+            [%observer i.t.wire ~]
+            %agent
+            [our.bowl app.observer]
+            %leave
+            ~
+        ==
+        :*  %pass
+            wire
+            %agent
+            [our.bowl app.observer]
+            %leave
+            ~
+        == 
+    ==
+  ::
+      %watch-ack
+    ?~  p.sign  [~ this]
+    ?+    wire  (on-agent:def wire sign)
+        [%thread-result @ ~]  [~ this]
+        [%observer @ ~]
+      =/  =time  (slav %da i.t.wire)
+      [~ this(observers (~(del by observers) time))]
+    ==
+  ::
+      %fact
+    ?+    wire  (on-agent:def wire sign)
```

Comment:
I think this would be cleaner to route on wire before sign

Reply:
Done.